test/fuzz/CMakeLists.txt

```diff
@@ -53,11 +87,25 @@ function(create_fuzz_test)
     "SOURCES;LIBRARIES"
     ${ARGN}
   )
-  message(STATUS "Creating fuzz test ${FUZZ_PREFIX}_fuzzer")
   add_executable(${FUZZ_PREFIX}_fuzzer ${FUZZ_SOURCES})
+  # XXX: The call produces both test and target
+  # <${FUZZ_PREFIX}-deps> as a side effect.
+  _add_test_suite_target(${FUZZ_PREFIX}
+    LABELS ${TEST_SUITE_NAME}
+    DEPENDS ${FUZZ_LIBRARIES} ${FUZZ_PREFIX}_fuzzer
+  )
   # Use PUBLIC to force 'fuzzer_config' for all dependent targets.
   target_link_libraries(${FUZZ_PREFIX}_fuzzer PUBLIC ${FUZZ_LIBRARIES})
   set(FUZZ_TEST_TARGETS ${FUZZ_TEST_TARGETS} ${FUZZ_PREFIX}_fuzzer PARENT_SCOPE)
+  BuildLibFuzzerOptions(libFuzzer_opts ${FUZZ_PREFIX})
+  set(test_title test/fuzz/${FUZZ_PREFIX}_test)
+  add_test(NAME ${test_title}
+           COMMAND $<TARGET_FILE:${FUZZ_PREFIX}_fuzzer> ${libFuzzer_opts}
+  )
```

Comment:
This generates the following test command:
```sh
1730: Test command: /home/burii/reviews/tarantool/ligurio/test/fuzz/csv_fuzzer "-max_total_time=60 -mutate_depth=20 -print_final_stats=1 -print_pcs=1 -reduce_inputs=1 -reload=1 -report_slow_units=5 -use_value_profile=1 -workers= -runs=$\{RUNS:-10\} -artifact_prefix=csv_"
```

As you can see, it lost the `corpus_dir` (<test/static/corpus/csv/>). I suppose it is better to set it separately (not as a flag options).

Also, these double quotes alow to parse the binary only the first option IINM:
```sh
test/fuzz/csv_fuzzer "-runs=1 -help=1"
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 1507478884
INFO: Loaded 1 modules   (116 inline 8-bit counters): 116 [0x55acb5879ba0, 0x55acb5879c14), 
INFO: Loaded 1 PC tables (116 PCs): 116 [0x55acb5879c18,0x55acb587a358), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2      INITED cov: 25 ft: 26 corp: 1/1b exec/s: 0 rss: 28Mb
#2      DONE   cov: 25 ft: 26 corp: 1/1b lim: 4 exec/s: 0 rss: 28Mb

test/fuzz/csv_fuzzer "-help=1 -runs=1"
# Help message output
```

In this case, all other flags are lost.

---

May be it is better to avoid a separate function for forming the options and just list them here.
Empty variable `${DICT_OPT}` or `${CORPUS_PATH}` will be just ignored.

Reply:
Updated:

```diff
diff --git a/test/fuzz/CMakeLists.txt b/test/fuzz/CMakeLists.txt
index 1aa7d9cda7..033ba22e7b 100644
--- a/test/fuzz/CMakeLists.txt
+++ b/test/fuzz/CMakeLists.txt
@@ -49,36 +49,21 @@ set(FUZZ_TEST_TARGETS "")
 
 set(DEFAULT_RUNS_NUMBER 10)
 
-string(JOIN " " DEFAULT_LIBFUZZER_OPTS
-    -max_total_time=60
-    -mutate_depth=20
-    -print_final_stats=1
-    -print_pcs=1
-    -reduce_inputs=1
-    -reload=1
-    -report_slow_units=5
-    -use_value_profile=1
-    -workers=${CMAKE_BUILD_PARALLEL_LEVEL}
-    -runs=$\\{RUNS:-${DEFAULT_RUNS_NUMBER}\\}
+list(APPEND LIBFUZZER_OPTS
+  -mutate_depth=20
+  -print_final_stats=1
+  -print_pcs=1
+  -reduce_inputs=1
+  -reload=1
+  -report_slow_units=5
+  -seed=${RANDOM_SEED}
+  -use_value_profile=1
+  -workers=${CMAKE_BUILD_PARALLEL_LEVEL}
+  # Shell parameter expansion,
+  # https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html
+  -runs=\${RUNS:-${DEFAULT_RUNS_NUMBER}}
 )
 
-# The function builds options for a fuzzing test.
-# The function sets the given variable in a parent scope to a
-# string with command-line options for specified test.
-function(BuildLibFuzzerOptions varname fuzz_name)
-  set(opts ${DEFAULT_LIBFUZZER_OPTS})
-  set(dict_path ${PROJECT_SOURCE_DIR}/test/static/${fuzz_name}.dict)
-  if(EXISTS ${dict_path})
-    set(opts "${opts} -dict=${dict_path}")
-  endif()
-  set(corpus_dir "${PROJECT_SOURCE_DIR}/test/static/corpus/${fuzz_name}")
-  if(EXISTS ${corpus_path})
-    set(opts "${opts} ${corpus_path}")
-  endif()
-  set(opts "${opts} -artifact_prefix=${fuzz_name}_")
-  set(${varname} ${opts} PARENT_SCOPE)
-endfunction()
-
 message(STATUS "Add test suite ${TEST_SUITE_NAME}")
 function(create_fuzz_test)
   cmake_parse_arguments(
@@ -92,19 +77,26 @@ function(create_fuzz_test)
   # Use PUBLIC to force 'fuzzer_config' for all dependent targets.
   target_link_libraries(${FUZZ_PREFIX}_fuzzer PUBLIC ${FUZZ_LIBRARIES})
   set(FUZZ_TEST_TARGETS ${FUZZ_TEST_TARGETS} ${FUZZ_PREFIX}_fuzzer PARENT_SCOPE)
-  BuildLibFuzzerOptions(libFuzzer_opts ${FUZZ_PREFIX})
+  set(LIBFUZZER_OPTS_EFFECTIVE ${LIBFUZZER_OPTS})
+  set(dict_path ${PROJECT_SOURCE_DIR}/test/static/${FUZZ_PREFIX}.dict)
+  if(EXISTS ${dict_path})
+    list(APPEND LIBFUZZER_OPTS_EFFECTIVE "-dict=${dict_path}")
+  endif()
   set(test_title test/fuzz/${FUZZ_PREFIX}_test)
   add_test(NAME ${test_title}
-           COMMAND $<TARGET_FILE:${FUZZ_PREFIX}_fuzzer> ${libFuzzer_opts}
+           COMMAND $<TARGET_FILE:${FUZZ_PREFIX}_fuzzer>
+                   ${LIBFUZZER_OPTS_EFFECTIVE}
+                   -artifact_prefix=${FUZZ_PREFIX}_
+                   ${PROJECT_SOURCE_DIR}/test/static/corpus/${FUZZ_PREFIX}
   )
   set_tests_properties(${test_title} PROPERTIES
```